core/classes/Endpoints/Endpoints.php

```diff
@@ -28,55 +33,162 @@ public function add(EndpointBase $endpoint): void {
+                    }
+
+                    $reflection = new ReflectionMethod($endpoint, 'execute');
+                    if ($reflection->getNumberOfParameters() !== (count($vars) + 1)) {
+                        throw new InvalidArgumentException("Endpoint's 'execute()' method must take " . (count($vars) + 1) . " arguments.");
+                    }
+
+                    $endpoint->execute(
+                        $api,
+                        ...array_map(function ($type, $value) use ($api) {
+                            return $this::transform($api, $type, $value);
+                        }, array_keys($vars), $vars)
+                    );
                     return;
                 }
-
             }
         }
 
         if ($matched_endpoint !== null) {
-            $api->throwError(3, $api->getLanguage()->get('api', 'invalid_api_method'), "The $route endpoint only accepts " . join(', ', $available_methods) . ", $method was used.", 405);
+            $api->throwError(3, $api->getLanguage()->get('api', 'invalid_api_method'), "The $route endpoint only accepts " . implode(', ', $available_methods) . ", $method was used.", 405);
             return;
         }
 
         $api->throwError(3, $api->getLanguage()->get('api', 'invalid_api_method'), 'If you are seeing this while in a browser, this does not mean your API is not functioning!', 404);
     }
 
     /**
-     * Get all registered Endpoints
+     * Determine if an Endpoint matches a route.
+     * If it does, return an array of variables to pass to the endpoint.
      *
-     * @return EndpointBase[] All endpoints.
+     * @param EndpointBase $endpoint Endpoint to attempt to match.
+     * @param string $route Route to match.
+     * @param Nameless2API $api Instance of API instance to provide the endpoint.
+     * @return array|false Array of variables to pass to the endpoint, or false if the route does not match.
      */
-    public function getAll(): iterable {
-        return $this->_endpoints;
+    private function matchRoute(EndpointBase $endpoint, string $route, Nameless2API $api) {
+        if (in_array($route, $endpoint->getRouteAliases(), true)) {
+            $error = str_replace(['{x}', '{y}'], [$route, $endpoint->getRoute()], $api->getLanguage()->get('api', 'route_alias_used'));
+
+            Log::getInstance()->log(Log::Action('api/route_alias_used'), $error);
+
+            $api->throwError(3, $api->getLanguage()->get('api', 'route_alias_used'), $error);
+            die();
+        }
+
+        $endpoint_parts = explode('/', $endpoint->getRoute());
+        $endpoint_vars = [];
+
+        $route_parts = explode('/', $route);
+        $route_vars = [];
+
+        $i = 0;
+        // first, find any variables (e.g. {user}) in the endpoint's route
+        // we save them to an array with their index, so we can reference them later
+        foreach ($endpoint_parts as $part) {
```

Comment:
As we've already discussed, this could be
```php
foreach ($endpoint_parts as $i => $part) {
```
to save manually creating + updating the $i var